Feature/core

cbintz opened the pull request:
As the start of my test project, I created a POJO person class, made a singleton person list object, and started to work on the person factory.

Commit:
person class added

Commit:
Gender class added

Commit:
singleton and factory work

Commit:
new static methods to generate people of different ages and genders

Commit:
edited static methods in factory to adjust birthdate and age based on lowAge & highAge arguments. made age and birthdate nullable in person POJO class and in factory.

Commit:
add numOfPeople amount of random types of people to person list in factory.

gregoryfein commented on CorinneTest/EmptyClass.cs:
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+namespace CorinneTest
+{
+    public class EmptyClass
```
Why do we have an empty class?  Why is it also called "EmptyClass"?

cbintz replied:
thought that I deleted this - now deleted.

gregoryfein commented on CorinneTest/Factory.cs:
```diff
@@ -0,0 +1,158 @@
+            int i = 0;
+            while (i < numOfPeople)
+            {
+                int type = rnd.Next(1, 9);
```
Use a switch statement?

gregoryfein replied:
Also, logically, I'm not sure this would compile.  "newPerson" is defined within the if() or else if().  HOw would "list.Add" know about the "newPerson" obj?

cbintz replied:
changed to use switch statement and defined newPerson outside of the switch statement

gregoryfein commented on CorinneTest/Factory.cs:
```diff
@@ -0,0 +1,158 @@
+            DateTime start = new DateTime((2018 - highAge), 5, 29);
+            DateTime minDate = new DateTime((2018 - lowAge), 5, 29);
+            int range = (minDate - start).Days;
+            return start.AddDays(rnd.Next(range));
```
cool logic.

as a slight C#-ism, the right-hand side of a declaration will infer to the compiler the left-hand type.  while its clean to specify DateTime or int on the left, you could just say "var" since the right-hand side is explicit that it will be a DateTime or int or whatever

cbintz replied:
changed to var - good to know.

gregoryfein commented on CorinneTest/Factory.cs:
```diff
@@ -0,0 +1,158 @@
+            if (lowAge >= 18) { return null; }
+            if (highAge < 18) { DateTime maxDate = new DateTime((2018 - highAge), 5, 29); }
+            else { DateTime maxDate = new DateTime(2000, 5, 29); }
+            int range = (DateTime.Today-maxDate).Days;
```
scoping issue.  MaxDate is defined within the logic blocks, it would inaccessible after the if {} else {}

define it above the if-else

cbintz replied:
changed

gregoryfein commented on CorinneTest/Factory.cs:
```diff
@@ -0,0 +1,158 @@
+        {
+            if (highAge < 6) { return null; }
+            if (lowAge > 65) { DateTime minDate = new DateTime((2018 - lowAge), 5, 29);}
+            else { DateTime minDate = new DateTime(1953, 5, 29); }
```
scoping issue

cbintz replied:
changed

gregoryfein commented on CorinneTest/Factory.cs:
```diff
@@ -0,0 +1,158 @@
+            return start.AddDays(rnd.Next(range));
+        }
+
+        public static int? GenAdultAge(int highAge)
```
hell yeah nullable

gregoryfein commented on CorinneTest/Factory.cs:
```diff
@@ -0,0 +1,158 @@
+            int age = rnd.Next(lowAge, highAge);
+            return age;
+        }
+
```
maybe make #region #endregion for similar methods (like all GenAge stuff in a region)?

cbintz replied:
good idea - did this or GenAge, GenBirthDate, and MakePerson

gregoryfein commented on CorinneTest/IPersonFactory.cs:
```diff
@@ -0,0 +1,8 @@
+﻿using System;
```
are you actually using these interfaces?

cbintz replied:
no - deleting them.

gregoryfein commented on CorinneTest/IPerson.cs:
```diff
@@ -0,0 +1,12 @@
+﻿using System;
```
are you actually using these interfaces?

cbintz replied:
not necessary for these projects - deleting IPerson and IPersonFactory interfaces

gregoryfein commented on CorinneTest/Person.cs:
```diff
@@ -0,0 +1,46 @@
+            set {
+                if (value < 1) value = 1;
+                else if (value > 100) value = 100;
+                this.age = value;
```
I'd make the getter NOT nullable.  And when you return this.age set it to an actual number (maybe -1?) if it was null

gregoryfein replied:
`get {
if(this.age == null)
this.Age= 1;
return this.age;
}`

notice we use Age instead of age in setting so that it goes through your set logic and not just setting the variable

cbintz replied:
so should it be "public int Age" instead of "public int? Age" ? if i do that then I get an error of converting an int? to an int

cbintz replied:
@gregoryfein

gregoryfein commented on CorinneTest/Person.cs:
```diff
@@ -0,0 +1,46 @@
+        private DateTime? birthDate;
+        public DateTime? BirthDate
+        {
+            get => this.birthDate; 
```
good job not having a setter here.  only a getter.  setting is done privately inside in construcotr.

gregoryfein commented on CorinneTest/concreteProductPerson.cs:
```diff
@@ -0,0 +1,10 @@
+﻿using System;
```
are we using this?

cbintz replied:
no - i thought i had deleted it earlier but really just removed it from the project

nnieslan commented on CorinneTest/IPerson.cs:
```diff
@@ -0,0 +1,12 @@
+    {
+        string Name();
+        int Age();
+        //Person.Gender PersonGender();
```
probably need to replace this with the new `Gender` that isn't nested in `Person`

cbintz replied:
yes that makes sense. I'm deleting IPerson and IPersonFactory interfaces because they aren't necessary for this project.

gregoryfein reviewed (changes requested):
Please address comments

nnieslan commented on CorinneTest/IPersonFactory.cs:
```diff
@@ -0,0 +1,8 @@
+{
+    interface IPersonFactory
+    {
+        Person CreatePerson(string myName, int myAge, Gender myGender, DateTime myBirthDate);
```
this should return the interface `IPerson` to allow different implementations don't you think?

cbintz replied:
yeah that would make sense for projects with different implementations. greg said i won't be doing that for this project so i'm deleting IPerson and IPersonFactory interfaces

nnieslan commented on CorinneTest/Person.cs:
```diff
@@ -0,0 +1,46 @@
+﻿using System;
+namespace CorinneTest
+{
+    public class Person 
```
You created an interface `IPerson` but didn't implement that interface here on the concrete class.

cbintz replied:
Yeah I added those interfaces when I was searching how to build a factory but they won't be necessary for this project.

nnieslan commented on CorinneTest/Person.cs:
```diff
@@ -0,0 +1,46 @@
+            get => this.birthDate; 
+        }
+
+        public Person(string myName, int? myAge, Gender myGender, DateTime? myBirthDate)
```
Generally constructors should be the first thing in your class after private field declaration.

nnieslan replied:
Also, should you do any parameter validation here?  What happens if `myName` is `null` or empty?  Is that ok?

gregoryfein replied:
As nick said, coding cleanliness usually puts things in this order:

- Variables
- Constructor
- Instance Methods
- Then whatever else

cbintz replied:
so Person constructor after private variable declarations and then the getter/setter methods?

cbintz replied:
@nnieslan

nnieslan commented on CorinneTest/PersonDataSource.cs:
```diff
@@ -0,0 +1,35 @@
+
+        private PersonDataSource()
+        {
+            this.People = new List<Person>();
```
Probably want to use `List<IPerson>`

nnieslan commented on CorinneTest/PersonDataSource.cs:
```diff
@@ -0,0 +1,35 @@
+        #endregion
+
+        #region Variables
+        public List<Person> People { get; set; }
```
Probably want to use List<IPerson>

gregoryfein replied:
Either List without the <> (at which point the T-Type would be the base "object" class ala Java.Lang.Object) or make it List< IPerson > (assuming you implement the interface)

nnieslan replied:
using the non-generic one (`List` with no `<>`) is very undesirable don't you think @gregoryfein ? You'd need to cast each object as you go or use reflection to determine types and you lose all of your type safety.

gregoryfein replied:
@nnieslan agreed.  but that was going to be part of my exercise about providing the right generic and how generic do you want to be ;)

cbintz replied:
@nnieslan what do you mean use List? like not have a List of type Person?

gregoryfein commented on CorinneTest/Person.cs:
```diff
@@ -0,0 +1,46 @@
+            }
+        }
+
+        private Gender gender;
```
using the word private isn't necessary since it is the default access of a variable.  i know that its a bit of a java holdover in your coding style, but the Static Analysis (an algorithm that checks your code and grades it) would complain about using "private"

cbintz replied:
got it

nnieslan commented on CorinneTest/Factory.cs:
```diff
@@ -0,0 +1,158 @@
+
+namespace CorinneTest
+{
+    class Factory : IPersonFactory
```
I'd say that any function in this class that isn't part of the `IPersonFactory` interface signature should be marked `private` or `protected` to keep the scope of the functions within the implementation of this class.  Otherwise, you could imagine another person on your team taking a dependency on the function even though it's not on the interface and that would be bad for maintainability and break the contract.

nnieslan commented on CorinneTest/Factory.cs:
```diff
@@ -0,0 +1,158 @@
+        {
+            var list = new List<Person>();
+            int i = 0;
+            while (i < numOfPeople)
```
where do you increment `i` to allow breaking out of the loop?

cbintz replied:
oops - changed this to increment i outside of the switch statement.

nnieslan reviewed (commented):
I added a number of comments that I saw as I quickly scanned this.  good progress!

gregoryfein commented on CorinneTest/Person.cs:
```diff
@@ -0,0 +1,46 @@
+            set { this.name = value; }
+        }
+
+        private int? age;
```
side note, you could completely remove int age and always return a computed value in get{} only from birthDate and have no set.

Commit:
case statments

Commit:
Using the Factory in the iOS bootup

AppDelegate.cs in iOS is using the factory to just pump out a bunch of peeps.  useful for actually making sure code builds, executes as intended, and can be inspected via debugging

Commit:
removed redundant private specifications

Commit:
Merge branch 'feature/core' of https://github.com/myStrength/CorinneTest into feature/core

# Conflicts:
#	CorinneTest/Factory.cs

Commit:
cleanup?

Commit:
fix get method for age in person class

Commit:
fixed order of constructor/variables in class

gregoryfein commented on CorinneTest/CorinneTest.sln:
```diff
@@ -0,0 +1,17 @@
+
```
we probably didnt want this committed but ill just take it out after the PR is merged

gregoryfein replied:
certain files are in the ".gitignore" which means they wont be tracked for modifications. this is useful for generated stuff like /bin/ or /obj/ . In java parlance youd want to save all of your .Java classes, but not your .Class compiled stuff since thats not useful to anyone

gregoryfein commented on CorinneTest/Factory.cs:
```diff
@@ -0,0 +1,195 @@
+            while (i < numOfPeople)
+            {
+                Person newPerson = null;
+                switch (rnd.Next(1, 9))
```
good switch here

gregoryfein commented on CorinneTest/Factory.cs:
```diff
@@ -0,0 +1,195 @@
+        {
+            DateTime minDate;
+            if (highAge < 6) { return null; }
+            if (lowAge > 65) { minDate = new DateTime((2018 - lowAge), 5, 29); }
```
we could make an enum that stores all of the high/lows for each range. not necessary but jus sptiballing.

gregoryfein replied:
similarly, 5/29 is arbitrary here but its fine for this context

gregoryfein commented on CorinneTest/Factory.cs:
```diff
@@ -0,0 +1,195 @@
+        {
+            if (highAge < 65)
+                return null;
+            int age = rnd.Next(65, highAge);
```
considering you have GenAge(int low, int high), all of these other methods could return that with unique params

gregoryfein replied:
i.e. GetSeniorAge could

'return GenAge(65, highAge);'

gregoryfein commented on CorinneTest/Factory.cs:
```diff
@@ -0,0 +1,195 @@
+
+#endregion
+
+    }
```
theres some white space we could clean up in here but not a big deal

cbintz commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Mobile Ramp-Up: Core - POCO](https://trello.com/c/HBOsEegd/35-mobile-ramp-up-core-poco)

cbintz commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Mobile Ramp-Up: Core - POCO](https://trello.com/c/HBOsEegd/35-mobile-ramp-up-core-poco)

cbintz commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Mobile Ramp-Up: Core - Factory](https://trello.com/c/wAAz6Box/37-mobile-ramp-up-core-factory)

cbintz commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Mobile Ramp-Up: Core - Singleton](https://trello.com/c/pbf2L7ck/36-mobile-ramp-up-core-singleton)

cbintz commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Mobile Ramp-Up: Core - Singleton](https://trello.com/c/pbf2L7ck/36-mobile-ramp-up-core-singleton)
